File: app/src/main/java/omniex/nl/omniex/ui/app/main/MainMenuActivity.java
package omniex.nl.omniex.ui.app.main;

import android.app.AlertDialog;
import android.widget.ImageView;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.ViewById;

import java.util.ArrayList;
import java.util.List;

import nl.omniex.omniexshopping.R;
import nl.omniex.omniexshopping.data.model.MenuItem;
import nl.omniex.omniexshopping.data.model.cart.Cart;
import nl.omniex.omniexshopping.data.model.cart.CartItemDelete;
import nl.omniex.omniexshopping.data.model.cart.CartQuantitySetter;
import nl.omniex.omniexshopping.ui.app.about.AboutFragment_;
import nl.omniex.omniexshopping.ui.app.auth.StartActivity_;
import nl.omniex.omniexshopping.ui.app.categories.CategoriesFragment_;
import nl.omniex.omniexshopping.ui.app.main.home.HomeFragment_;
import nl.omniex.omniexshopping.ui.app.newsletter.NewsletterFragment_;
import nl.omniex.omniexshopping.ui.app.order.OrderActivity_;
import nl.omniex.omniexshopping.ui.app.product.list.ProductsListFragment_;
import nl.omniex.omniexshopping.ui.app.profile.ProfileFragment_;
import nl.omniex.omniexshopping.ui.base.menu.BaseMenuActivity;
import nl.omniex.omniexshopping.ui.base.menu.MenuAdapter;
import nl.omniex.omniexshopping.ui.views.dialogs.cart.CartDialog;
import nl.omniex.omniexshopping.ui.views.dialogs.cart.CartDialog_;
import nl.omniex.omniexshopping.utils.SharedPrefUtils;

@EActivity(R.layout.activity_main_menu)
public class MainMenuActivity extends BaseMenuActivity<MenuAdapter, MainMenuView, MainMenuPresenter> implements MainMenuView, CartDialog.OnUpdateItemQuantityListener, CartDialog.OnMakeOrderClickListener {

    @ViewById(R.id.duo_view_menu_background)
    ImageView mDuoDrawerLayout;

    private CartDialog mCartDialog;
    private boolean mIsCartOpen;

    @AfterViews
    void initHomeFragment() {
        mDuoDrawerLayout.setBackgroundResource(R.color.colorPrimaryDark);
        goToFragment(HomeFragment_.builder().build(), true, "");
    }

    @Override
    protected void onFirstCreate() {
        super.onFirstCreate();
        setToolbar();
    }

    @Override
    protected int getOverlayRes() {
        return R.id.menu_overlay;
    }

    @Override
    public int getToolbarId() {
        return 0;
    }

    @Override
    public int getDuoDrawerLayout() {
        return R.id.home_drawer;
    }

    @Override
    public int getStartIndex() {
        return 0;
    }

    @Override
    public void onOptionClicked(int position, Object objectClicked) {
        if (objectClicked != null) {
            MenuItem menuItem = (MenuItem) objectClicked;
            switch (position) {
                case 0:
                    goToFragment(HomeFragment_.builder().build(), false, "");
                    break;
                case 1:
                    goToFragment(ProductsListFragment_.builder().mIsBestSellersList(true).build(), false, "");
                    break;
                case 2:
                    goToFragment(CategoriesFragment_.builder().build(), false, "");
                    break;
                case 3:
                    handleProfileClick();
                    break;
                case 4:
                    handleNewsletterClick();
                    break;
                case 5:
                    goToFragment(AboutFragment_.builder().build(), false, "");
                    break;
                case 6:
                    handleLogoutClick();
                    break;
            }
        }
    }

    @Override
    public MenuAdapter createAdapter() {
        List<MenuItem> menuItems = new ArrayList<>();

        if (SharedPrefUtils.isUserLogged()) {
            menuItems.add(new MenuItem("Featured Products"));
            menuItems.add(new MenuItem("Bestsellers"));
            menuItems.add(new MenuItem("Products"));
            menuItems.add(new MenuItem("Profile"));
            menuItems.add(new MenuItem("Newsletter"));
            menuItems.add(new MenuItem("About Omniex"));
            menuItems.add(new MenuItem("Logout"));
        } else {
            menuItems.add(new MenuItem("Featured Products"));
            menuItems.add(new MenuItem("Bestsellers"));
            menuItems.add(new MenuItem("Products"));
            menuItems.add(new MenuItem("Profile"));
            menuItems.add(new MenuItem("Newsletter"));
            menuItems.add(new MenuItem("About Omniex"));
            menuItems.add(new MenuItem("Login"));
        }
        return new MenuAdapter(menuItems);
    }

    private void handleLogoutClick() {
        if (SharedPrefUtils.isUserLogged()) {
            getPresenter().logout();
        } else {
            finish();
            StartActivity_.intent(this).start();
        }
    }

    private void handleProfileClick() {
        if (SharedPrefUtils.isUserLogged()) {
            goToFragment(ProfileFragment_.builder().build(), false, "");
        } else {
            showGuestWarnDialog();
        }
    }

    private void handleNewsletterClick() {
        if (SharedPrefUtils.isUserLogged()) {
            goToFragment(NewsletterFragment_.builder().build(), false, "");
        } else {
            showGuestWarnDialog();
        }
    }

    public void showGuestWarnDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setCancelable(false)
                .setMessage("Please log in to get access to this function.")
                .setPositiveButton("Login",
                        ((dialog, which) -> {
                            finish();
                            dialog.dismiss();
                            StartActivity_.intent(this).start();
                        }))
                .setNegativeButton("Later", ((dialog, which) -> {
                    dialog.dismiss();
                }))
                .show();
    }

    @Override
    public void onBackPressed() {
        toggleMenu();
    }

    private void setToolbar() {
        getCustomToolbar()
                .setIconEnd(R.drawable.twotone_shopping_cart_black_36)
                .setIconEndClickListener(() -> getPresenter().getCart());
    }

    @Override
    public void onCartFetched(Cart cart) {
        if (SharedPrefUtils.isUserLogged()) {
            if (!mIsCartOpen) {
                mCartDialog = CartDialog_.builder().mCart(cart).build().setOnUpdateItemQuantityListener(this).setOnMakeOrderClickListener(this);
                mCartDialog.show(getSupportFragmentManager(), "dialog");
                mIsCartOpen = true;
            } else {
                mCartDialog.refreshCart(cart);
            }
        } else {
            showGuestWarnDialog();
        }
    }

    @Override
    public void onMakeOrderClick() {
        OrderActivity_.intent(this).start();
    }

    @Override
    public void onCartDismiss() {
        mIsCartOpen = false;
    }

    @Override
    public void onUpdateQuantity(String productKey, int quantity) {
        getPresenter().updateCartQuantity(new CartQuantitySetter(productKey, quantity));
    }

    @Override
    public void onRemoveCartItem(CartItemDelete cartItemDelete) {
        getPresenter().deleteCartItem(cartItemDelete);
    }

    @Override
    public void onCartItemQuantityUpdated() {
        getPresenter().getCart();
    }

    @Override
    public void onCartEmpty() {
        if (SharedPrefUtils.isUserLogged()) {
            if (mCartDialog != null && mCartDialog.isVisible())
                mCartDialog.dismiss();
            AlertDialog.Builder builder = new AlertDialog.Builder(this).setMessage("Cart is empty.").setPositiveButton("OK", ((dialog, which) -> dialog.dismiss()));
            builder.show();
        } else {
            showGuestWarnDialog();
        }
    }

    @Override
    public void onLogoutSuccess() {
        finishAffinity();
        StartActivity_.intent(this).start();
    }

    @Override
    public void startLoading() {
        showProgressBar();
    }

    @Override
    public void stopLoading() {
        hideProgressBar();
    }
}

File: app/src/main/java/omniex/nl/omniex/ui/app/main/MainMenuActivity.kt
package omniex.nl.omniex.ui.app.main

import android.support.v7.app.AlertDialog
import android.widget.ImageView
import omniex.nl.omniex.R
import omniex.nl.omniex.data.model.MenuItem
import omniex.nl.omniex.data.model.cart.Cart
import omniex.nl.omniex.data.model.cart.CartItemDelete
import omniex.nl.omniex.data.model.cart.CartQuantitySetter
import omniex.nl.omniex.ui.app.about.AboutFragment_
import omniex.nl.omniex.ui.app.auth.StartActivity_
import omniex.nl.omniex.ui.app.categories.CategoriesFragment_
import omniex.nl.omniex.ui.app.main.home.HomeFragment_
import omniex.nl.omniex.ui.app.newsletter.NewsletterFragment_
import omniex.nl.omniex.ui.app.order.OrderActivity_
import omniex.nl.omniex.ui.app.product.list.ProductsListFragment_
import omniex.nl.omniex.ui.app.profile.ProfileFragment_
import omniex.nl.omniex.ui.base.menu.BaseMenuActivity
import omniex.nl.omniex.ui.base.menu.MenuAdapter
import omniex.nl.omniex.ui.base.menu.MenuHelper
import omniex.nl.omniex.ui.views.dialogs.cart.CartDialog
import omniex.nl.omniex.ui.views.dialogs.cart.CartDialog_
import omniex.nl.omniex.ui.views.toolbar.CustomToolbar
import omniex.nl.omniex.utils.SharedPrefUtils
import org.androidannotations.annotations.AfterViews
import org.androidannotations.annotations.EActivity
import org.androidannotations.annotations.ViewById
import java.util.*
import javax.inject.Inject


@EActivity(R.layout.activity_main_menu)
open class MainMenuActivity : BaseMenuActivity<MenuAdapter, MainMenuView, MainMenuPresenter>(), MainMenuView, MenuHelper.MenuBase<MenuAdapter>, CartDialog.OnUpdateItemQuantityListener, CartDialog.OnMakeOrderClickListener {

    @ViewById(R.id.duo_view_menu_background)
    lateinit var mDuo: ImageView

       private var mCartDialog: CartDialog? = null
    private var mIsCartOpen: Boolean = false

    override val overlayRes: Int
        get() = R.id.menu_overlay

    override fun toolbarId(): Int = 0

    override fun duoDrawerLayout(): Int = R.id.home_drawer

    override fun startIndex(): Int = 0

    @Inject
    lateinit var mSharedPrefUtils: SharedPrefUtils

    @AfterViews
    internal fun initHomeFragment() {
        mDuo!!.setBackgroundResource(R.color.colorPrimaryDark)
        goToFragment(HomeFragment_.builder().build(), true, "")
    }

    override fun onFirstCreate() {
        super.onFirstCreate()
        setToolbar()
    }

    override fun onOptionClicked(position: Int, objectClicked: Any) {
        if (objectClicked != null) {
            when (position) {
                0 -> goToFragment(HomeFragment_.builder().build(), false, "")
                1 -> goToFragment(ProductsListFragment_.builder().mIsBestSellersList(true).build(), false, "")
                2 -> goToFragment(CategoriesFragment_.builder().build(), false, "")
                3 -> handleProfileClick()
                4 -> handleNewsletterClick()
                5 -> goToFragment(AboutFragment_.builder().build(), false, "")
                6 -> handleLogoutClick()
            }
        }
    }

    override fun createAdapter(): MenuAdapter {
        val menuItems = ArrayList<MenuItem>()
        val isTrue= true

        if (isTrue) {
            menuItems.add(MenuItem("Featured Products"))
            menuItems.add(MenuItem("Bestsellers"))
            menuItems.add(MenuItem("Products"))
            menuItems.add(MenuItem("Profile"))
            menuItems.add(MenuItem("Newsletter"))
            menuItems.add(MenuItem("About Omniex"))
            menuItems.add(MenuItem("Logout"))
        } else {
            menuItems.add(MenuItem("Featured Products"))
            menuItems.add(MenuItem("Bestsellers"))
            menuItems.add(MenuItem("Products"))
            menuItems.add(MenuItem("Profile"))
            menuItems.add(MenuItem("Newsletter"))
            menuItems.add(MenuItem("About Omniex"))
            menuItems.add(MenuItem("Login"))
        }
        return MenuAdapter(menuItems)
    }
//
    private fun handleLogoutClick() {
        if (mSharedPrefUtils.isUserLogged()) {
            getPresenter().logout()
        } else {
            finish()
            StartActivity_.intent(this).start()
        }
    }

    private fun handleProfileClick() {
        if (mSharedPrefUtils.isUserLogged()) {
            goToFragment(ProfileFragment_.builder().build(), false, "")
        } else {
            showGuestWarnDialog()
        }
    }

    private fun handleNewsletterClick() {
        if (mSharedPrefUtils.isUserLogged()) {
            goToFragment(NewsletterFragment_.builder().build(), false, "")
        } else {
            showGuestWarnDialog()
        }
    }

    fun showGuestWarnDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setCancelable(false)
                .setMessage("Please log in to get access to this function.")
                .setPositiveButton("Login"
                ) { dialog, which ->
                    finish()
                    dialog.dismiss()
                    StartActivity_.intent(this).start()
                }
                .setNegativeButton("Later") { dialog, which -> dialog.dismiss() }
                .show()
    }

    override fun onBackPressed() {
        toggleMenu()
    }

    fun setToolbar() {
        customToolbar
                .setIconEnd(R.drawable.twotone_shopping_cart_black_36)
                .setIconEndClickListener(object : CustomToolbar.IconEndClickListner{
                    override fun onIconEndClick() {
                        getPresenter().getCart()
                    }
                })
    }

    override fun onCartFetched(cart: Cart) {
        if (mSharedPrefUtils.isUserLogged()) {
            if (!mIsCartOpen) {
                mCartDialog = CartDialog_.builder().mCart(cart).build().setOnUpdateItemQuantityListener(this).setOnMakeOrderClickListener(this)
                mCartDialog!!.show(getSupportFragmentManager(), "dialog")
                mIsCartOpen = true
            } else {
                mCartDialog!!.refreshCart(cart)
            }
        } else {
            showGuestWarnDialog()
        }
    }

   override fun onMakeOrderClick() {
        OrderActivity_.intent(this).start()
    }

    override fun onCartDismiss() {
        mIsCartOpen = false
    }

    override fun onUpdateQuantity(productKey: String, quantity: Int) {
        getPresenter().updateCartQuantity(CartQuantitySetter(productKey, quantity))
    }

    override fun onRemoveCartItem(cartItemDelete: CartItemDelete) {
        getPresenter().deleteCartItem(cartItemDelete)
    }

    override fun onCartItemQuantityUpdated() {
//        getPresenter().getCart()
    }

    override fun onCartEmpty() {
        if (mSharedPrefUtils.isUserLogged()) {
            if (mCartDialog != null && mCartDialog!!.isVisible())
                mCartDialog!!.dismiss()
            val builder = AlertDialog.Builder(this).setMessage("Cart is empty.").setPositiveButton("OK") { dialog, which -> dialog.dismiss() }
            builder.show()
        } else {
            showGuestWarnDialog()
        }
    }

    override fun onLogoutSuccess() {
        finishAffinity()
        StartActivity_.intent(this).start()
    }
//
    override fun startLoading() {
        showProgressBar()
    }

    override fun stopLoading() {
        hideProgressBar()
    }
}
